### PR TITLE
Create Order - Fix remove cart rule when amount is not reached

### LINKED ISSUE #21812: BO - Order page - The voucher line is not removed when the Order's total < Minimum amount for cart rule
#### Describe the bug

In the BO > Order page, the voucher line is not removed when the Order's total < Minimum amount for cart rule.

#### Expected behavior

The voucher line should be removed

#### Steps to Reproduce

1. Go to BO > Catalog > Discounts page
2. Create a cart rule:
2.1 Informations: With a code
2.2 Conditions: Minimum amount = 50euro, tax excluded, shipping excluded
2.3 Actions: Apply a discount amount = 1 euro tax included
image
3. Go to BO => Orders => Add new order page for "John Doe"
4. add some product to the order
5. Add demo_6 to the cart with quantity =1 (29 euro tax excluded)
6. Add demo_21 to the cart with quantity = 1 (35 euro tax excluded)
7. Search the discount and apply it => ok
8. remove the product demo_21
9. see error => the voucher line is not removed
10. In the summary block, the voucher is applied => NOK

image

PS1768 => ok

#### Additional information

* PrestaShop version: 1770rc1build2
* PHP version: 7.2

### PR DESCRIPTION
**Pull request by sowbiba**
**Breaking change** : ContextStateManager is added as dependency of the RemoveProductFromCartHandler to set the correct cart, customer, currency, shop when updating the cart

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 1.7.7.x
| Description?  | When a cart rule with minimum amount is added to an order, if the order doesn't reach the required amount (because of a product removal), the cart rule was not removed. Technically, the problem was that the cart was not added to the context, so context->cart was empty
| Type?         | bug fix
| Category?     | BO
| BC breaks?    | yes
| Deprecations? | no
| Fixed ticket? | Fixes #21812
| How to test?  | See __Steps to Reproduce__ of the issue

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/22077)

### CODE REVIEW

**Review by jolelievre — COMMENTED** [review]
I know you expected this, but can you add Behat tests please? ^^

**Review by Progi1984 — CHANGES_REQUESTED** [review]
Need behat tests

**Review by Progi1984 — DISMISSED** [review]

File: tests/Integration/Behaviour/Features/Scenario/Cart/Adding/CartRule/minimum_amount_cartrule.feature [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    And I specify its status as enabled
+    And I specify its "code" as "CART_RULE_MIN_AMOUNT"
+    And its minimum purchase amount in currency "USD" is "50"
+    And its minimum purchase amount is tax excluded
+    And its minimum purchase amount is shipping included
+    And it gives a reduction amount of "2" in currency "USD" which is tax included and applies to order without shipping
+    And I save it
+    When I add 1 products "product1" to the cart "dummy_custom_cart"
+    When I add 1 products "product4" to the cart "dummy_custom_cart"
+    And cart "dummy_custom_cart" should contain 2 products
```
Comment (atomiix) [inline tests/Integration/Behaviour/Features/Scenario/Cart/Adding/CartRule/minimum_amount_cartrule.feature]:
```suggestion
    Then cart "dummy_custom_cart" should contain 2 products
```

File: tests/Integration/Behaviour/Features/Scenario/Cart/Adding/CartRule/minimum_amount_cartrule.feature [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    When I use a voucher "CART_RULE_MIN_AMOUNT" on the cart "dummy_custom_cart"
+    And reduction value of voucher "CART_RULE_MIN_AMOUNT" in cart "dummy_custom_cart" should be "2"
```
Comment (atomiix) [inline tests/Integration/Behaviour/Features/Scenario/Cart/Adding/CartRule/minimum_amount_cartrule.feature]:
```suggestion
    Then reduction value of voucher "CART_RULE_MIN_AMOUNT" in cart "dummy_custom_cart" should be "2"
```

File: tests/Integration/Behaviour/Features/Scenario/Cart/Adding/CartRule/minimum_amount_cartrule.feature [hunk 3]
```diff
@@ -0,0 +1,45 @@
+# ./vendor/bin/behat -c tests/Integration/Behaviour/behat.yml -s cart --tags cart-minimum-amount-cart-rule
+@reset-database-before-feature
+@cart-minimum-amount-cart-rule
+Feature: Add cart rule in cart
+  As a customer
+  I must be able to correctly add cart rules in my cart
+
+  Background:
+    And there is customer "testCustomer" with email "pub@prestashop.com"
```
Comment (atomiix) [inline tests/Integration/Behaviour/Features/Scenario/Cart/Adding/CartRule/minimum_amount_cartrule.feature]:
```suggestion
    Given there is customer "testCustomer" with email "pub@prestashop.com"
```

File: tests/Integration/Behaviour/Features/Scenario/Cart/Adding/CartRule/minimum_amount_cartrule.feature [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    When I add 1 products "product1" to the cart "dummy_custom_cart"
+    When I add 1 products "product4" to the cart "dummy_custom_cart"
+    And cart "dummy_custom_cart" should contain 2 products
+    When I use a voucher "CART_RULE_MIN_AMOUNT" on the cart "dummy_custom_cart"
+    And reduction value of voucher "CART_RULE_MIN_AMOUNT" in cart "dummy_custom_cart" should be "2"
+    And cart "dummy_custom_cart" total with tax included should be "$53.38"
+    When I delete product "product1" from cart "dummy_custom_cart"
+    And cart "dummy_custom_cart" should contain 1 products
+    And cart "dummy_custom_cart" total with tax included should be "$35.57"
+    Then voucher "CART_RULE_MIN_AMOUNT" should not be applied to cart "dummy_custom_cart"
```
Comment (atomiix) [inline tests/Integration/Behaviour/Features/Scenario/Cart/Adding/CartRule/minimum_amount_cartrule.feature]:
```suggestion
    Then cart "dummy_custom_cart" should contain 1 products
    And cart "dummy_custom_cart" total with tax included should be "$35.57"
    And voucher "CART_RULE_MIN_AMOUNT" should not be applied to cart "dummy_custom_cart"
```

**Review by Progi1984 — APPROVED** [review]

File: src/Adapter/Cart/CommandHandler/RemoveProductFromCartHandler.php [hunk 5]
```diff
@@ -39,21 +43,47 @@
...
 {
+    /**
+     * @var ContextStateManager
+     */
+    private $contextStateManager;
+
+    /**
+     * @param ContextStateManager $contextStateManager
+     */
+    public function __construct(
```
Comment (atomiix) [inline src/Adapter/Cart/CommandHandler/RemoveProductFromCartHandler.php:54]: I thinks that's a BC break

**Review by atomiix — APPROVED** [review]
I approve it but don't forget to document the BC Break

File: src/Adapter/Cart/CommandHandler/RemoveProductFromCartHandler.php (hunk 5 above)
Comment (matthieu-rolland) [reply src/Adapter/Cart/CommandHandler/RemoveProductFromCartHandler.php:54]: It looks like this class didn't exist before 1.7.7 it should be fine 👍

Comment (matthieu-rolland) [reply src/Adapter/Cart/CommandHandler/RemoveProductFromCartHandler.php:54]: oops, no, that's a 1.7.7.1 PR, it's a bc break then...

**Review by matthieu-rolland — APPROVED** [review]

File: src/Adapter/Cart/CommandHandler/RemoveProductFromCartHandler.php (hunk 5 above)
Comment (sowbiba) [reply src/Adapter/Cart/CommandHandler/RemoveProductFromCartHandler.php:54]: :+1: Done

**Comment by SD1982** [comment]
LGTM Thanks @sowbiba !!
https://drive.google.com/file/d/1i5QKYGNsAdU6QFWX0q9Q1S1zFP_CffnH/view?usp=sharing

**Comment by Progi1984** [comment]
Thanks @sowbiba & @SD1982

**Pull request merged**